aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -2804,17 +2804,36 @@ static void linalg_lstsq_out_info(
   // now the actual call that computes the result in-place (apply_lstsq)
   at::_lstsq_helper_(solution, rank, singular_values, infos, input_working_copy, rcond, driver);
 
+  // residuals are available only if m > n and drivers other than gelsy used
   if (m > n && driver != "gelsy") {
-    // LAPACK stores residuals data for postprocessing in rows n:(m-n)
-    auto raw_residuals = solution.narrow(/*dim=*/-2, /*start=*/n, /*length*/m - n);
-    if (raw_residuals.is_complex()) {
-      // in-place abs is not supported for complex tensors
-      raw_residuals = raw_residuals.abs();
-      raw_residuals.pow_(2);
-    } else {
-      raw_residuals.abs_().pow_(2);
+    // if the driver is gelss or gelsd then the residuals are available only if rank == n
+    bool compute_residuals = true;
+    if (driver != "gels") {
+      if (input.dim() == 2) {
+        compute_residuals = (rank.item().toInt() == n);
+      } else {
+        // if any of the computed ranks in equal to n then do the computation of residuals
+        compute_residuals = at::any(rank == n).item().toBool();
+      }
+    }
+    if (compute_residuals) {
+      // LAPACK stores residuals data for postprocessing in rows n:(m-n)
+      auto raw_residuals = solution.narrow(/*dim=*/-2, /*start=*/n, /*length*/m - n);
+      if (raw_residuals.is_complex()) {
+        // in-place abs is not supported for complex tensors
+        raw_residuals = raw_residuals.abs();
+        raw_residuals.pow_(2);
+      } else {
+        raw_residuals.abs_().pow_(2);
+      }
```

Comment:
This block of code could be replaced with
```
raw_residuals.mul_(raw_residuals.conj())
```
which should work for both real and complex.